### PR TITLE
[RFR] Introduce <PasswordInput>

### PR DESCRIPTION
**Pull request by Kmaschta**
This PR introduce a new component in `ra-ui-materialui`, `<PasswordInput>`.
It is very similar to a `<TextInput type="password">`, but it adds a button to toggle visibility.

I found that it's a common need in our client projects, and it can be a great accessibility addition.

That being said, I'm willing to create a third-party library with this component if you think that this component doesn't have its place in `ra-ui-materialui`.

Also, I have another proposal: apply this toggle button directly on the `TextInput` if the type is password.

## TODO
- [x] Implement the component
- [x] Update the documentation
- [x] Use it in the demo

## Previews

image

image

image

### CODE REVIEW

File: docs/Inputs.md [hunk 1]
```diff
@@ -629,6 +629,29 @@ You can customize the `step` props (which defaults to "any"):
...
+
+```jsx
+import { PasswordInput } from 'react-admin';
+
+<PasswordInput source="password" />
+```
+
+![Password Input](./img/password-input.png)
+
+It is possible to change the default behavior and display the value by default via the `initiallyHidden` prop:
```
Comment (fzaninotto) [inline docs/Inputs.md]: initiallyVisible

File: examples/demo/src/visitors/VisitorEdit.js [hunk 2]
```diff
@@ -45,7 +46,11 @@ const VisitorEdit = ({ classes, ...props }) => (
                     fullWidth={true}
                     formClassName={classes.email}
                 />
-                <DateInput source="birthday" />
+                <DateInput source="birthday" formClassName={classes.birthday} />
+                <PasswordInput
```
Comment (fzaninotto) [inline examples/demo/src/visitors/VisitorEdit.js]: No, the password should never be edited like this in an admin. An admin should not be able to see a user password. In fact, consider that the password is stored hashed, so it's not editable. The API should never expose the password, and the admin should use a password field to create a new password for an existing user.

File: examples/data-generator/src/customers.js [hunk 3]
```diff
@@ -11,11 +11,14 @@ export default (db, { serializeDate }) =>
         const last_name = name.lastName();
         const email = internet.email(first_name, last_name);
         const birthday = has_ordered ? date.past(60) : null;
+        const password = internet.password();
```
Comment (fzaninotto) [inline examples/data-generator/src/customers.js]: no, the API should only store a password hash.

File: examples/demo/src/visitors/VisitorCreate.js [hunk 4]
```diff
@@ -44,7 +47,11 @@ const VisitorCreate = ({ classes, ...props }) => (
                     fullWidth={true}
                     formClassName={classes.email}
                 />
-                <DateInput source="birthday" />
+                <DateInput source="birthday" formClassName={classes.birthday} />
+                <PasswordInput
```
Comment (fzaninotto) [inline examples/demo/src/visitors/VisitorCreate.js]: We need to simulate that the clear password will be hashed before it's stored. Unfortunately we don't have API controllers for that...

File: packages/ra-ui-materialui/src/Link.js [hunk 5]
```diff
@@ -4,12 +4,13 @@ import classNames from 'classnames';
 import { Link as RRLink } from 'react-router-dom';
 import { withStyles, createStyles } from '@material-ui/core/styles';
 
-const styles = theme => createStyles({
-    link: {
-        textDecoration: 'none',
-        color: theme.palette.primary.main,
-    },
-});
+const styles = theme =>
```
Comment (fzaninotto) [inline packages/ra-ui-materialui/src/Link.js]: weird change. Can you revert this file?

File: packages/ra-ui-materialui/src/auth/Login.js [hunk 6]
```diff
@@ -15,30 +15,31 @@ import defaultTheme from '../defaultTheme';
...
-    avatar: {
-        margin: '1em',
-        display: 'flex',
-        justifyContent: 'center',
-    },
-    icon: {
-        backgroundColor: theme.palette.secondary[500],
-    },
-});
+const styles = theme =>
```
Comment (fzaninotto) [inline packages/ra-ui-materialui/src/auth/Login.js]: you changed nothing, can you revert this file?

File: packages/ra-ui-materialui/src/auth/LoginForm.js [hunk 7]
```diff
@@ -10,17 +10,18 @@ import CircularProgress from '@material-ui/core/CircularProgress';
...
-        padding: '0 1em 1em 1em',
-    },
-    input: {
-        marginTop: '1em',
-    },
-    button: {
-        width: '100%',
-    },
-});
+const styles = () =>
```
Comment (fzaninotto) [inline packages/ra-ui-materialui/src/auth/LoginForm.js]: you changed nothing. Can you revert this file?

File: packages/ra-ui-materialui/src/auth/Logout.js [hunk 8]
```diff
@@ -8,17 +8,18 @@ import ExitIcon from '@material-ui/icons/PowerSettingsNew';
...
-    },
-    iconMenuPaddingStyle: {
-        paddingRight: '1.2em',
-    },
-    iconPaddingStyle: {
-        paddingRight: theme.spacing.unit,
-    },
-});
+const styles = theme =>
+    createStyles({
```
Comment (fzaninotto) [inline packages/ra-ui-materialui/src/auth/Logout.js]: You changed nothing. Can you revert this file?

File: packages/ra-ui-materialui/src/detail/TabbedShowLayout.js [hunk 9]
```diff
@@ -101,23 +106,21 @@ export class TabbedShowLayout extends Component {
                 )}
                 <Divider />
                 <CardContentInner>
-                    {Children.map(
```
Comment (fzaninotto) [inline packages/ra-ui-materialui/src/detail/TabbedShowLayout.js]: can you revert this change?

File: examples/demo/src/visitors/VisitorCreate.js [hunk 10]
```diff
@@ -28,29 +31,20 @@ const VisitorCreate = ({ classes, ...props }) => (
...
                 <TextInput
                     type="email"
                     source="email"
                     validation={{ email: true }}
                     fullWidth={true}
                     formClassName={classes.email}
                 />
-                <DateInput source="birthday" />
+                <DateInput source="birthday" formClassName={classes.birthday} />
+                <PasswordInput source="password" formClassName={classes.password} />
```
Comment (fzaninotto) [inline examples/demo/src/visitors/VisitorCreate.js]: Since this is a demo, we should showcase how to hav two password fields and a validator that makes sure the two are correct

File: examples/demo/src/visitors/VisitorEdit.js [hunk 11]
```diff
@@ -12,7 +12,7 @@ import {
     ReferenceManyField,
     TabbedForm,
     TextField,
-    TextInput,
```
Comment (fzaninotto) [inline examples/demo/src/visitors/VisitorEdit.js]: I see only coding style changes in this file, you didn't add a way to update the password?

File: packages/ra-ui-materialui/src/input/PasswordInput.js [hunk 12]
```diff
@@ -0,0 +1,59 @@
+import React, { Component } from 'react';
```
Comment (fzaninotto) [inline packages/ra-ui-materialui/src/input/PasswordInput.js]: Since it's a new component, I invite you to write it directly in TypeScript.

**Comment by fzaninotto** [comment]
Any chance you can update that one soon?

**Comment by Kmaschta** [comment]
Yeah, I think I'll just do it in plain JavaScript since none of the inputs are in TS and that I need to use a dependency.

**Comment by fzaninotto** [comment]
OK, let's forget the TypeScript requirement

**Comment by Kmaschta** [comment]
OK, I replaced the direct password edition by a tab with two fields: password and confirm password, on both Create and Edit views.
But if I hash the password by changing the data provider, I fear that someone will use it as an example and it's not really a good practice.

Since the demo API is a false in-memory API with only purpose to show react admin features, are we forced to handle that password hashing?

**Comment by Luwangel** [comment]
We use this version in one of our project. It's not TypeScript but there are hooks :smile: 

``` js
import React, { useState } from 'react';
import PropTypes from 'prop-types';
import { TextInput } from 'react-admin';

import IconButton from '@material-ui/core/IconButton';
import InputAdornment from '@material-ui/core/InputAdornment';
import Visibility from '@material-ui/icons/Visibility';
import VisibilityOff from '@material-ui/icons/VisibilityOff';

const PasswordInput = ({ defaultVisible, ...otherProps }) => {
    const [visible, setVisible] = useState(defaultVisible);

    return (
        <TextInput
            {...otherProps}
            type={visible ? 'text' : 'password'}
            InputProps={{
                endAdornment: (
                    <InputAdornment position="end">
                        <IconButton onClick={() => setVisible(!visible)}>
                            {visible ? <Visibility /> : <VisibilityOff />}
                        </IconButton>
                    </InputAdornment>
                ),
            }}
        />
    );
};

PasswordInput.propTypes = {
    defaultVisible: PropTypes.bool,
};

PasswordInput.defaultProps = {
    defaultVisible: false,
};

export default PasswordInput;
```

**Comment by Kmaschta** [comment]
Is it worth it to rebase this one?
Or we don't need a `<PasswordInput>` in the core?
Or should I rewrite it entirely for the v3?

**Comment by fzaninotto** [comment]
Yes, it's still worth rebasing! No need to rewrite it with hooks, but you're welcome to do so.

**Comment by Kmaschta** [comment]
Rebased on `next`

File: packages/ra-ui-materialui/src/input/PasswordInput.js [hunk 13]
```diff
@@ -0,0 +1,50 @@
...
+PasswordInput.propTypes = {
+    translate: PropTypes.func.isRequired,
+    initiallyVisible: PropTypes.bool,
+};
+
+PasswordInput.defaultProps = {
+    initiallyVisible: false,
+};
+
+export default compose(
```
Comment (djhi) [inline packages/ra-ui-materialui/src/input/PasswordInput.js]: Please migrate this component to hooks:
`useTranslate` and `useInput` from `ra-core`

**Comment by Kmaschta** [comment]
Done :ok_hand:

File: examples/demo/src/i18n/en.js [hunk 14]
```diff
@@ -51,10 +51,16 @@ export default {
...
                 reviews: 'Reviews',
                 stats: 'Stats',
+                password: 'Password',
+                change_password: 'Change Password',
             },
             page: {
                 delete: 'Delete Customer',
             },
+            errors: {
+                password_mismatch:
```
Comment (djhi) [inline examples/demo/src/i18n/en.js]: This message is missing in the french translations

File: examples/demo/src/visitors/VisitorCreate.js [hunk 15]
```diff
@@ -64,9 +84,24 @@ const VisitorCreate = props => {
...
+                        required
+                    />
+                </FormTab>
             </TabbedForm>
         </Create>
     );
 };
 
-export default VisitorCreate;
+export default withTranslation(VisitorCreate);
```
Comment (djhi) [inline examples/demo/src/visitors/VisitorCreate.js]: Use the `useTranslate` hook

File: examples/demo/src/visitors/VisitorEdit.js [hunk 16]
```diff
@@ -123,9 +126,22 @@ const VisitorEdit = props => {
...
+                        formClassName={classes.confirm_password}
+                    />
+                </FormTab>
             </TabbedForm>
         </Edit>
     );
 };
 
-export default VisitorEdit;
+export default withTranslation(VisitorEdit);
```
Comment (djhi) [inline examples/demo/src/visitors/VisitorEdit.js]: Same, `useTranslate`

File: packages/ra-ui-materialui/src/input/PasswordInput.tsx [hunk 17]
```diff
@@ -0,0 +1,54 @@
...
+    return (
+        <TextInput
+            {...rest}
+            type={visible ? 'text' : 'password'}
+            InputProps={{
+                endAdornment: (
+                    <InputAdornment position="end">
+                        <IconButton
+                            aria-label={translate(
+                                'ra.input.password.toggle_visibility'
```
Comment (djhi) [inline packages/ra-ui-materialui/src/input/PasswordInput.tsx]: I'm not fond of toggle. I'd rather have two explicit messages, one for each state

File: packages/ra-ui-materialui/src/input/PasswordInput.tsx [hunk 18]
```diff
@@ -0,0 +1,54 @@
...
+                        </IconButton>
+                    </InputAdornment>
+                ),
+            }}
+        />
+    );
+};
+
+PasswordInput.propTypes = {
+    initiallyVisible: PropTypes.bool,
```
Comment (ashim-k-saha) [inline packages/ra-ui-materialui/src/input/PasswordInput.tsx]: `<PasswordInput />` should accept one more prop, i.e, `autocomplete` with default value `current-password`.  Ref #3666

**Comment by fzaninotto** [comment]
Needs rebase

File: examples/demo/src/i18n/en.js (hunk 14 above)
Comment (Kmaschta) [reply examples/demo/src/i18n/en.js]: Done now :+1:

File: packages/ra-ui-materialui/src/input/PasswordInput.tsx (hunk 18 above)
Comment (Kmaschta) [reply packages/ra-ui-materialui/src/input/PasswordInput.tsx]: It is possible to do so by adding the prop `inputProps={{ autocomplete: 'current-password' }}`

Comment (Kmaschta) [reply packages/ra-ui-materialui/src/input/PasswordInput.tsx]: I added a tip about that on the documentation

**Comment by fzaninotto** [comment]
Excellent, thanks!
